clang/test/SemaSYCL/unique_stable_name.cpp

```diff
@@ -0,0 +1,119 @@
+// RUN: %clang_cc1 %s -fsyntax-only -ast-dump -fsycl-is-device -triple spir64 -verify
+
+// expected-error 4 @+2{{kernel instantiation changes the result of an evaluated __builtin_unique_stable_name}}
+template <typename KernelName, typename KernelType>
+[[clang::sycl_kernel]] void kernel_single_task(KernelType kernelFunc) {
+  kernelFunc();
+}
+template <template <typename> typename Outer, typename Inner>
+struct S {
+  void operator()() const;
+};
+
+template <typename Ty>
+struct T {};
+
+template <typename Func>
+void kernel3func(const Func &F3) {
+  // expected-note@+1{{__builtin_unique_stable_name evaluated here}}
+  constexpr const char *F3_output = __builtin_unique_stable_name(F3);
+  // expected-error@+1{{error: kernel instantiation changes the result of an evaluated __builtin_unique_stable_name}}
+  kernel_single_task<class kernel3>(F3);
+}
+
+template <typename Func>
+void kernel4func(const Func &F4) {
+  // expected-note@+1{{__builtin_unique_stable_name evaluated here}}
+  constexpr const char *F4_output = __builtin_unique_stable_name(F4);
+  // expected-error@+1{{error: kernel instantiation changes the result of an evaluated __builtin_unique_stable_name}}
```

Comment:
This error surprises me -- something odd seems to be going on here. CC @erichkeane

Reply:
It shouldn't :)  If "Func" is a lambda, it is required to name the new lambda on line 29.  So the name would change once line 29 happens.

Reply:
Ohhhh, @schittir, you should be sure to add comments to this test case explaining what's going on there, otherwise we'll look at this test a few years from now and be very confused.

Reply:
@erichkeane - did you mean that this error shouldn't happen?

Reply:
No, this error should happen.